types/src/eth.rs

```diff
@@ -0,0 +1,274 @@
+    #[serde(rename = "type")]
+    pub type_: u8,
+    pub nonce: u64,
+    pub gas: u64,
+    #[serde(rename = "gasPrice")]
+    pub gas_price: U256,
+    pub from: Address,
+    pub to: Option<Address>,
+    #[serde(rename = "chainId")]
+    pub chain_id: U256,
+    pub value: U256,
+    pub data: Bytes,
+    #[serde(rename = "isCreate")]
+    pub is_create: bool,
+    pub v: U64,
+    pub r: U256,
+    pub s: U256,
+}
+
+impl TransactionTrace {
+    pub fn to_eth_tx(
+        &self,
+        block_hash: Option<H256>,
+        block_number: Option<U64>,
+        transaction_index: Option<U64>,
+    ) -> Transaction {
+        Transaction {
+            hash: Default::default(),
+            nonce: U256::from(self.nonce),
+            block_hash,
+            block_number,
+            transaction_index,
+            from: self.from,
+            to: self.to,
+            value: self.value,
+            gas_price: Some(self.gas_price),
+            gas: U256::from(self.gas),
+            input: self.data.clone(),
+            v: self.v,
+            r: self.r,
+            s: self.s,
+            transaction_type: None,
+            access_list: None,
+            max_priority_fee_per_gas: None,
+            max_fee_per_gas: None,
+            chain_id: Some(self.chain_id),
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct ExecutionResult {
+    pub gas: u64,
+    pub failed: bool,
+    #[serde(rename = "returnValue", default)]
+    pub return_value: String,
+    pub from: Option<AccountProofWrapper>,
+    pub to: Option<AccountProofWrapper>,
+    #[serde(rename = "codeHash")]
+    pub code_hash: Option<Hash>,
+    #[serde(rename = "byteCode")]
+    pub byte_code: Option<String>,
+    #[serde(rename = "structLogs")]
+    pub exec_steps: Vec<ExecStep>,
+}
+
+impl ExecutionResult {
+    pub fn to_geth_exec_trace(&self) -> GethExecTrace {
+        let mut struct_logs = Vec::new();
+        for exec_step in &self.exec_steps {
+            let step = exec_step.to_geth_exec_step();
+            struct_logs.push(step)
+        }
+        fix_geth_trace_memory_size(&mut struct_logs);
+        GethExecTrace {
+            gas: Gas(self.gas),
+            failed: self.failed,
+            return_value: self.return_value.clone(),
+            struct_logs,
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct ExecStep {
+    pub pc: u64,
+    pub op: OpcodeId,
+    pub gas: u64,
+    #[serde(rename = "gasCost")]
+    pub gas_cost: u64,
+    #[serde(default)]
+    pub refund: u64,
+    pub depth: isize,
+    pub error: Option<String>,
+    pub stack: Option<Vec<Word>>,
+    pub memory: Option<Vec<Word>>,
+    pub storage: Option<HashMap<Word, Word>>,
+    #[serde(rename = "extraData")]
+    pub extra_data: Option<ExtraData>,
+}
+
+impl ExecStep {
+    #[allow(dead_code)]
+    fn to_geth_exec_step(&self) -> GethExecStep {
+        let stack = if let Some(stack) = self.stack.clone() {
```

Comment:
More concise way to write this would be:

```rust
let stack = self.stack.and_then(Stack::from).unwrap_or_else(Stack::new);
```

Reply:
This goes for all other bits of code that look like it

Reply:
Tried, but seems cannot compile.
```
 let stack = e.stack.and_then(Stack::from).unwrap_or_else(Stack::new);
                                 ^^^^^^^^ expected struct `Stack`, found enum `std::option::Option`

```

Reply:
I think maybe `map_or_else` is better